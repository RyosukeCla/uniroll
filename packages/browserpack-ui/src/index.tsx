export { EnvContext, useEnv } from "./components/contexts";
export { App } from "./components/App";
