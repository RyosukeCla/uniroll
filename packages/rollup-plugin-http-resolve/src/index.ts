import { Plugin } from "rollup";
import path from "path";

export type ImportMaps = {
  imports: { [k: string]: string };
};

function isHttpProtocol(id: string) {
  return id.startsWith("http://") || id.startsWith("https://");
}

const DEBUG = false;
const log = (...args: any) => DEBUG && console.log(...args);

type HttpResolveOptions = {
  cache?: any;
  fetcher?: (url: string) => Promise<string>;
  onRequest?: (url: string) => void;
  onUseCache?: (url: string) => void;
  transform?: (code: string) => string;
  fallback?: (
    id: string,
    importer: string,
    onWarn: (warn: any) => void
  ) => Promise<string | void> | void | string;
};
const defaultCache = new Map();
export const httpResolve = function httpResolve_({
  cache = defaultCache,
  onRequest,
  transform,
  onUseCache,
  fetcher,
  fallback,
}: HttpResolveOptions) {
  return {
    name: "http-resolve",
    async resolveId(id: string, importer: string) {
      log("[http-resolve:resolveId:enter]", id, "from", importer);
      // on network resolve
      if (importer && isHttpProtocol(importer)) {
        if (id.startsWith("https://")) {
          log("[http-reslove:end] return with https", id);
          return id;
        }
        const { pathname, protocol, host } = new URL(importer);
        if (id.startsWith("/")) {
          // /_/...
          log(
            "[http-reslove:end] return with host root",
            `${protocol}//${host}${id}`
          );
          return `${protocol}//${host}${id}`;
        } else if (id.startsWith(".")) {
          // ./xxx/yyy
          // relative path
          const resolvedPathname = path.join(path.dirname(pathname), id);
          const newId = `${protocol}//${host}${resolvedPathname}`;
          log("[http-resolve:end] return with relativePath", newId);
          return newId;
        }
      }
      if (fallback) {
        const fallbacked = await fallback(id, importer, this.warn);
        log("[http-resolve:end] use fallback to", id, "=>", fallbacked);
        if (fallbacked) {
          return fallbacked;
        }
      }
    },
    async load(id: string) {
      log("[http-resolve:load]", id);
      if (id === null) {
        this.warn("irregular missing id");
        return;
      }
      if (isHttpProtocol(id)) {
        const cached = await cache.get(id);
        if (cached) {
          onUseCache?.(id);
          return cached;
        }
        onRequest?.(id);
        if (fetcher) {
          let code = await fetcher(id);
          code = transform?.(code) ?? code;
          await cache.set(id, code);
          return code;
        } else {
          const res = await fetch(id);
          if (!res.ok) {
            throw res.statusText;
          }
          let code = await res.text();
          code = transform?.(code) ?? code;
          await cache.set(id, code);
          return code;
        }
      }
    },
  } as Plugin;
};

export function createImportMapsFallback({
  importmaps,
}: {
  importmaps?: ImportMaps;
}) {
  return (
    id: string,
    importer: string | void = undefined,
    warn: (warning: any) => any
  ): Promise<void | string> | void | string => {
    if (importer == null) {
      return;
    }
    if (id.startsWith("http")) {
      return;
    }
    if (id.startsWith(".")) {
      return;
    }

    const mapped = importmaps?.imports[id];
    if (mapped) {
      return mapped;
    }
    warn(`missed fallback to https://cdn.skypack.dev/${id}`);
    return `https://cdn.skypack.dev/${id}`;
  };
}

// export default httpResolve;
