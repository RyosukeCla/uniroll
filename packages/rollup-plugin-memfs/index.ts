import type { Plugin } from "rollup";
import type { IPromisesAPI } from "memfs/lib/promises";
import path from "path";

export type ResolverOptions = {
  extensions?: string[];
  searchDirectoryIndex?: boolean;
  searchDirectoryPackage?: boolean;
}

const DEFALUT_EXTENSIONS = [".ts", ".tsx", ".js", ".mjs", ".json"];
export const memfsPlugin = (fs: IPromisesAPI) => {
  const resolver = new Resolver(fs);
  return {
    name: "memfs",
    async resolveId(source: string, importer: string | undefined) {
      // console.log("[memfs.resolveId]", source, importer);

      // resolve relative path;
      if (importer && importer.startsWith("/") && source.startsWith(".")) {
        const fullpath = importer
          ? path.resolve(path.dirname(importer), source)
          : source;
        const reslovedWithExt = await resolver.resolveId(fullpath);
        return reslovedWithExt;
      }
      return source;
    },
    async load(id: string) {
      const m = await fs.readFile(id, "utf-8");
      return m;
    }
  } as Plugin;
};

class Resolver {
  private extensions: string[];
  private searchDirectoryIndex: boolean;
  private searchDirectoryPackage: boolean;
  constructor(
    private fs: IPromisesAPI,
    {
      extensions = DEFALUT_EXTENSIONS,
      searchDirectoryIndex = true,
      searchDirectoryPackage = false
    }: ResolverOptions = {}
  ) {
    this.extensions = extensions;
    // always include id itself first
    if (!this.extensions.includes("")) {
      this.extensions.unshift("");
    }
    this.searchDirectoryIndex = searchDirectoryIndex;
    this.searchDirectoryPackage = searchDirectoryPackage;
  }

  async exists(id: string): Promise<boolean> {
    return this.fs.access(id).then(() => true).catch(_err => false);
  }
  async resolveId(id: string): Promise<string | void> {
    // [id][ext]
    for (const ext of this.extensions) {
      const s = id + ext;
      if (await this.exists(s)) {
        return id + ext;
      }
    }

    for (const ext of this.extensions) {
      const s = path.join(id, "index" + ext);
      if (await this.exists(s)) {
        return id + ext;
      }
    }

    // [id]/index[ext]
    if (this.searchDirectoryIndex) {
      for (const ext of this.extensions) {
        const s = path.join(id, "index" + ext);
        if (await this.exists(s)) {
          return s;
        }
      }
    }
    // [id]/package.json#main
    if (this.searchDirectoryPackage) {
      const pkgPath = path.join(id, "package.json");
      if ((await this.exists(pkgPath))) {
        return;
      }
      const pkgStr = await this.fs.readFile(pkgPath);
      let json;
      try {
        json = JSON.parse(pkgStr.toString());
      } catch (err) {
        console.warn(err);
        // through parse error
      }
      if (json && json.module) {
        return json.module;
      }
      if (json && json.main) {
        return json.main;
      }
    }
  }
}
