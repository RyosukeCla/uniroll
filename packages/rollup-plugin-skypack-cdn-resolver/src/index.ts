import { Plugin } from "rollup";
import path from "path";
const SKYPACK_CDN_HOST = "https://cdn.skypack.dev";

export function skypackCDNResolver({
  cache = new Map(),
  onRequest,
  onUseCache,
  ignorePolyfill = false,
}: {
  cache?: any;
  ignorePolyfill?: boolean;
  onRequest?: (url: string) => void;
  onUseCache?: (url: string) => void;
}) {
  return {
    async resolveId(id: string, importer: string) {
      // console.log(id, importer);
      if (importer && importer.startsWith(SKYPACK_CDN_HOST)) {
        // load Skypack in Skypack
        if (id.startsWith(SKYPACK_CDN_HOST)) {
          return id;
        }

        const { pathname, protocol, host } = new URL(importer);
        if (id.startsWith("/")) {
          return `${protocol}//${host}${id}`;
        } else {
          const resolvedPathname = path.join(path.dirname(pathname), id);
          const newId = `${protocol}//${host}${resolvedPathname}`;
          return newId;
        }
      }
    },
    async load(id: string) {
      if (id.includes("@pika/polyfill")) {
        return `// ignored: ${id}`;
      }
      if (id.startsWith(SKYPACK_CDN_HOST)) {
        const cached = await cache.get(id);
        if (cached) {
          onUseCache && onUseCache(id);
          return cached;
        }
        onRequest && onRequest(id);
        const res = await fetch(id);
        if (!res.ok) {
          throw res.statusText;
        }
        const code = await res.text();
        await cache.set(id, code);
        return code;
      }
    },
  } as Plugin;
}
