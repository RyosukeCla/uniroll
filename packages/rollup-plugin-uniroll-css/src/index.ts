import { Options } from "./types";
import { Plugin } from "rollup";
import { transformCss } from "./transformCss";
import { transformWithAutoprefixer } from "./transformWithAutoprefixer";
import { wrapWithStyleInjector } from "./wrapWithInjector";

export const css = (options: Options = {}) => {
  return {
    async transform(code: string, id: string): Promise<string | void> {
      if (id.endsWith(".css")) {
        const css = await transformCss(code, options);
        const autoprefixedCss = await transformWithAutoprefixer(css);
        return wrapWithStyleInjector(autoprefixedCss);
      }
    }
  } as Plugin;
};
