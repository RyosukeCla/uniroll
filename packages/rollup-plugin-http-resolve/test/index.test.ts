import "isomorphic-unfetch";

import createFs, { IPromisesAPI } from "memfs/lib/promises";
import { httpResolve } from "../src/index";
import { memfsPlugin } from "rollup-plugin-memfs";

import { rollup } from "rollup";
import { Volume } from "memfs";
import ts from "typescript";
test("build with esm.sh", async () => {
  const vol = Volume.fromJSON({
    "/index.js": `
    import {h} from "https://esm.sh/preact";
    console.log(h);
    `,
  });

  const memfs = createFs(vol) as IPromisesAPI;
  const rolled = await rollup({
    input: "/index.js",
    plugins: [httpResolve(), memfsPlugin(memfs)],
  });
  const out = await rolled.generate({ format: "es" });
  const code = out.output[0].code;
  expect(code).toMatchSnapshot();
});

test("build with esm.sh", async () => {
  const vol = Volume.fromJSON({
    "/index.js": `
    import {h} from "https://esm.sh/preact";
    console.log(h);
    `,
  });

  const memfs = createFs(vol) as IPromisesAPI;
  const rolled = await rollup({
    input: "/index.js",
    plugins: [httpResolve(), memfsPlugin(memfs)],
  });
  const out = await rolled.generate({ format: "es" });
  const code = out.output[0].code;
  expect(code).toMatchSnapshot();
});

test("build with skypack", async () => {
  const vol = Volume.fromJSON({
    "/index.js": `
    import {h} from "https://cdn.skypack.dev/preact";
    console.log(h);
    `,
  });

  const memfs = createFs(vol) as IPromisesAPI;
  const rolled = await rollup({
    input: "/index.js",
    plugins: [httpResolve(), memfsPlugin(memfs)],
  });
  const out = await rolled.generate({ format: "es" });
  const code = out.output[0].code;
  expect(code).toMatchSnapshot();
});

test("build with fallback", async () => {
  const vol = Volume.fromJSON({
    "/index.js": `
    import {h} from "preact";
    console.log(h);
    `,
  });

  const memfs = createFs(vol) as IPromisesAPI;
  const rolled = await rollup({
    input: "/index.js",
    plugins: [
      httpResolve({
        fallback(id) {
          if (!id.startsWith(".")) {
            return `https://esm.sh/${id}`;
          }
        },
      }),
      memfsPlugin(memfs),
    ],
  });
  const out = await rolled.generate({ format: "es" });
  const code = out.output[0].code;
  expect(code).toMatchSnapshot();
});

test("build nested with skypack", async () => {
  const vol = Volume.fromJSON({
    "/index.js": `
    import { h } from "https://cdn.skypack.dev/preact@10.4.6";
    import { useEffect } from "https://cdn.skypack.dev/preact@10.4.6/hooks";
    function App() {
      useEffect(() => {
        console.log("effect");
      }, []);
      return h("div");
    }

    console.log(App);
    `,
  });

  const memfs = createFs(vol) as IPromisesAPI;

  const rolled = await rollup({
    input: "/index.js",
    plugins: [httpResolve(), memfsPlugin(memfs)],
  });
  const out = await rolled.generate({ format: "es" });
  const code = out.output[0].code;
  expect(code).toMatchSnapshot();
});

test("with transform", async () => {
  const vol = Volume.fromJSON({
    "/index.js": `
    import * as svelteInternal from "https://cdn.skypack.dev/svelte/internal";
    console.log(svelteInternal);
    `,
  });

  const memfs = createFs(vol) as IPromisesAPI;
  const rolled = await rollup({
    input: "/index.js",
    plugins: [
      httpResolve(),
      memfsPlugin(memfs),
      {
        name: "transform-cdn",
        transform(code, id) {
          if (id?.startsWith("https://")) {
            const out = ts.transpileModule(code, {
              compilerOptions: {
                module: ts.ModuleKind.ESNext,
                target: ts.ScriptTarget.ES5,
              },
            });
            return {
              code: out.outputText,
              map: out.sourceMapText,
            };
          }
        },
      },
    ],
  });
  const out = await rolled.generate({ format: "iife" });
  const code = out.output[0].code;
  expect(code).toContain("/** @class */");
});
