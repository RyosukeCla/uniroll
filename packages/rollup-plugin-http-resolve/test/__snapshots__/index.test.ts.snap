// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build nested with skypack 1`] = `
"var n, u, i, t, o, f = {}, e = [], c = /acit|ex(?:s|g|n|p|$)|rph|grid|ows|mnc|ntw|ine[ch]|zoo|^ord|itera/i;
function s(n2, l2) {
  for (var u2 in l2)
    n2[u2] = l2[u2];
  return n2;
}
function a(n2) {
  var l2 = n2.parentNode;
  l2 && l2.removeChild(n2);
}
function v(n2, l2, u2) {
  var i2, t2 = arguments, r2 = {};
  for (i2 in l2)
    i2 !== \\"key\\" && i2 !== \\"ref\\" && (r2[i2] = l2[i2]);
  if (arguments.length > 3)
    for (u2 = [u2], i2 = 3; i2 < arguments.length; i2++)
      u2.push(t2[i2]);
  if (u2 != null && (r2.children = u2), typeof n2 == \\"function\\" && n2.defaultProps != null)
    for (i2 in n2.defaultProps)
      r2[i2] === void 0 && (r2[i2] = n2.defaultProps[i2]);
  return h(n2, r2, l2 && l2.key, l2 && l2.ref, null);
}
function h(l2, u2, i2, t2, r2) {
  var o2 = {type: l2, props: u2, key: i2, ref: t2, __k: null, __: null, __b: 0, __e: null, __d: void 0, __c: null, constructor: void 0, __v: r2};
  return r2 == null && (o2.__v = o2), n.vnode && n.vnode(o2), o2;
}
function p(n2) {
  return n2.children;
}
function d(n2, l2) {
  this.props = n2, this.context = l2;
}
function _(n2, l2) {
  if (l2 == null)
    return n2.__ ? _(n2.__, n2.__.__k.indexOf(n2) + 1) : null;
  for (var u2; l2 < n2.__k.length; l2++)
    if ((u2 = n2.__k[l2]) != null && u2.__e != null)
      return u2.__e;
  return typeof n2.type == \\"function\\" ? _(n2) : null;
}
function k(n2) {
  var l2, u2;
  if ((n2 = n2.__) != null && n2.__c != null) {
    for (n2.__e = n2.__c.base = null, l2 = 0; l2 < n2.__k.length; l2++)
      if ((u2 = n2.__k[l2]) != null && u2.__e != null) {
        n2.__e = n2.__c.base = u2.__e;
        break;
      }
    return k(n2);
  }
}
function w(l2) {
  (!l2.__d && (l2.__d = true) && u.push(l2) && !m.__r++ || t !== n.debounceRendering) && ((t = n.debounceRendering) || i)(m);
}
function m() {
  for (var n2; m.__r = u.length; )
    n2 = u.sort(function(n3, l2) {
      return n3.__v.__b - l2.__v.__b;
    }), u = [], n2.some(function(n3) {
      var l2, u2, i2, t2, r2, o2, f2;
      n3.__d && (o2 = (r2 = (l2 = n3).__v).__e, (f2 = l2.__P) && (u2 = [], (i2 = s({}, r2)).__v = i2, t2 = T(f2, r2, i2, l2.__n, f2.ownerSVGElement !== void 0, null, u2, o2 == null ? _(r2) : o2), $(u2, r2), t2 != o2 && k(r2)));
    });
}
function g(n2, l2, u2, i2, t2, r2, o2, c2, s2, v2) {
  var y2, d2, k2, w2, m2, g2, b2, A2 = i2 && i2.__k || e, P2 = A2.length;
  for (s2 == f && (s2 = o2 != null ? o2[0] : P2 ? _(i2, 0) : null), u2.__k = [], y2 = 0; y2 < l2.length; y2++)
    if ((w2 = u2.__k[y2] = (w2 = l2[y2]) == null || typeof w2 == \\"boolean\\" ? null : typeof w2 == \\"string\\" || typeof w2 == \\"number\\" ? h(null, w2, null, null, w2) : Array.isArray(w2) ? h(p, {children: w2}, null, null, null) : w2.__e != null || w2.__c != null ? h(w2.type, w2.props, w2.key, null, w2.__v) : w2) != null) {
      if (w2.__ = u2, w2.__b = u2.__b + 1, (k2 = A2[y2]) === null || k2 && w2.key == k2.key && w2.type === k2.type)
        A2[y2] = void 0;
      else
        for (d2 = 0; d2 < P2; d2++) {
          if ((k2 = A2[d2]) && w2.key == k2.key && w2.type === k2.type) {
            A2[d2] = void 0;
            break;
          }
          k2 = null;
        }
      m2 = T(n2, w2, k2 = k2 || f, t2, r2, o2, c2, s2, v2), (d2 = w2.ref) && k2.ref != d2 && (b2 || (b2 = []), k2.ref && b2.push(k2.ref, null, w2), b2.push(d2, w2.__c || m2, w2)), m2 != null ? (g2 == null && (g2 = m2), s2 = x(n2, w2, k2, A2, o2, m2, s2), u2.type == \\"option\\" ? n2.value = \\"\\" : typeof u2.type == \\"function\\" && (u2.__d = s2)) : s2 && k2.__e == s2 && s2.parentNode != n2 && (s2 = _(k2));
    }
  if (u2.__e = g2, o2 != null && typeof u2.type != \\"function\\")
    for (y2 = o2.length; y2--; )
      o2[y2] != null && a(o2[y2]);
  for (y2 = P2; y2--; )
    A2[y2] != null && I(A2[y2], A2[y2]);
  if (b2)
    for (y2 = 0; y2 < b2.length; y2++)
      H(b2[y2], b2[++y2], b2[++y2]);
}
function x(n2, l2, u2, i2, t2, r2, o2) {
  var f2, e2, c2;
  if (l2.__d !== void 0)
    f2 = l2.__d, l2.__d = void 0;
  else if (t2 == u2 || r2 != o2 || r2.parentNode == null)
    n:
      if (o2 == null || o2.parentNode !== n2)
        n2.appendChild(r2), f2 = null;
      else {
        for (e2 = o2, c2 = 0; (e2 = e2.nextSibling) && c2 < i2.length; c2 += 2)
          if (e2 == r2)
            break n;
        n2.insertBefore(r2, o2), f2 = o2;
      }
  return f2 !== void 0 ? f2 : r2.nextSibling;
}
function A(n2, l2, u2, i2, t2) {
  var r2;
  for (r2 in u2)
    r2 === \\"children\\" || r2 === \\"key\\" || r2 in l2 || C(n2, r2, null, u2[r2], i2);
  for (r2 in l2)
    t2 && typeof l2[r2] != \\"function\\" || r2 === \\"children\\" || r2 === \\"key\\" || r2 === \\"value\\" || r2 === \\"checked\\" || u2[r2] === l2[r2] || C(n2, r2, l2[r2], u2[r2], i2);
}
function P(n2, l2, u2) {
  l2[0] === \\"-\\" ? n2.setProperty(l2, u2) : n2[l2] = typeof u2 == \\"number\\" && c.test(l2) === false ? u2 + \\"px\\" : u2 == null ? \\"\\" : u2;
}
function C(n2, l2, u2, i2, t2) {
  var r2, o2, f2, e2, c2;
  if (t2 ? l2 === \\"className\\" && (l2 = \\"class\\") : l2 === \\"class\\" && (l2 = \\"className\\"), l2 === \\"style\\")
    if (r2 = n2.style, typeof u2 == \\"string\\")
      r2.cssText = u2;
    else {
      if (typeof i2 == \\"string\\" && (r2.cssText = \\"\\", i2 = null), i2)
        for (e2 in i2)
          u2 && e2 in u2 || P(r2, e2, \\"\\");
      if (u2)
        for (c2 in u2)
          i2 && u2[c2] === i2[c2] || P(r2, c2, u2[c2]);
    }
  else
    l2[0] === \\"o\\" && l2[1] === \\"n\\" ? (o2 = l2 !== (l2 = l2.replace(/Capture$/, \\"\\")), f2 = l2.toLowerCase(), l2 = (f2 in n2 ? f2 : l2).slice(2), u2 ? (i2 || n2.addEventListener(l2, N, o2), (n2.l || (n2.l = {}))[l2] = u2) : n2.removeEventListener(l2, N, o2)) : l2 !== \\"list\\" && l2 !== \\"tagName\\" && l2 !== \\"form\\" && l2 !== \\"type\\" && l2 !== \\"size\\" && !t2 && l2 in n2 ? n2[l2] = u2 == null ? \\"\\" : u2 : typeof u2 != \\"function\\" && l2 !== \\"dangerouslySetInnerHTML\\" && (l2 !== (l2 = l2.replace(/^xlink:?/, \\"\\")) ? u2 == null || u2 === false ? n2.removeAttributeNS(\\"http://www.w3.org/1999/xlink\\", l2.toLowerCase()) : n2.setAttributeNS(\\"http://www.w3.org/1999/xlink\\", l2.toLowerCase(), u2) : u2 == null || u2 === false && !/^ar/.test(l2) ? n2.removeAttribute(l2) : n2.setAttribute(l2, u2));
}
function N(l2) {
  this.l[l2.type](n.event ? n.event(l2) : l2);
}
function z(n2, l2, u2) {
  var i2, t2;
  for (i2 = 0; i2 < n2.__k.length; i2++)
    (t2 = n2.__k[i2]) && (t2.__ = n2, t2.__e && (typeof t2.type == \\"function\\" && t2.__k.length > 1 && z(t2, l2, u2), l2 = x(u2, t2, t2, n2.__k, null, t2.__e, l2), typeof n2.type == \\"function\\" && (n2.__d = l2)));
}
function T(l2, u2, i2, t2, r2, o2, f2, e2, c2) {
  var a2, v2, h2, y2, _2, k2, w2, m2, b2, x2, A2, P2 = u2.type;
  if (u2.constructor !== void 0)
    return null;
  (a2 = n.__b) && a2(u2);
  try {
    n:
      if (typeof P2 == \\"function\\") {
        if (m2 = u2.props, b2 = (a2 = P2.contextType) && t2[a2.__c], x2 = a2 ? b2 ? b2.props.value : a2.__ : t2, i2.__c ? w2 = (v2 = u2.__c = i2.__c).__ = v2.__E : (\\"prototype\\" in P2 && P2.prototype.render ? u2.__c = v2 = new P2(m2, x2) : (u2.__c = v2 = new d(m2, x2), v2.constructor = P2, v2.render = L), b2 && b2.sub(v2), v2.props = m2, v2.state || (v2.state = {}), v2.context = x2, v2.__n = t2, h2 = v2.__d = true, v2.__h = []), v2.__s == null && (v2.__s = v2.state), P2.getDerivedStateFromProps != null && (v2.__s == v2.state && (v2.__s = s({}, v2.__s)), s(v2.__s, P2.getDerivedStateFromProps(m2, v2.__s))), y2 = v2.props, _2 = v2.state, h2)
          P2.getDerivedStateFromProps == null && v2.componentWillMount != null && v2.componentWillMount(), v2.componentDidMount != null && v2.__h.push(v2.componentDidMount);
        else {
          if (P2.getDerivedStateFromProps == null && m2 !== y2 && v2.componentWillReceiveProps != null && v2.componentWillReceiveProps(m2, x2), !v2.__e && v2.shouldComponentUpdate != null && v2.shouldComponentUpdate(m2, v2.__s, x2) === false || u2.__v === i2.__v) {
            v2.props = m2, v2.state = v2.__s, u2.__v !== i2.__v && (v2.__d = false), v2.__v = u2, u2.__e = i2.__e, u2.__k = i2.__k, v2.__h.length && f2.push(v2), z(u2, e2, l2);
            break n;
          }
          v2.componentWillUpdate != null && v2.componentWillUpdate(m2, v2.__s, x2), v2.componentDidUpdate != null && v2.__h.push(function() {
            v2.componentDidUpdate(y2, _2, k2);
          });
        }
        v2.context = x2, v2.props = m2, v2.state = v2.__s, (a2 = n.__r) && a2(u2), v2.__d = false, v2.__v = u2, v2.__P = l2, a2 = v2.render(v2.props, v2.state, v2.context), v2.state = v2.__s, v2.getChildContext != null && (t2 = s(s({}, t2), v2.getChildContext())), h2 || v2.getSnapshotBeforeUpdate == null || (k2 = v2.getSnapshotBeforeUpdate(y2, _2)), A2 = a2 != null && a2.type == p && a2.key == null ? a2.props.children : a2, g(l2, Array.isArray(A2) ? A2 : [A2], u2, i2, t2, r2, o2, f2, e2, c2), v2.base = u2.__e, v2.__h.length && f2.push(v2), w2 && (v2.__E = v2.__ = null), v2.__e = false;
      } else
        o2 == null && u2.__v === i2.__v ? (u2.__k = i2.__k, u2.__e = i2.__e) : u2.__e = j(i2.__e, u2, i2, t2, r2, o2, f2, c2);
    (a2 = n.diffed) && a2(u2);
  } catch (l3) {
    u2.__v = null, n.__e(l3, u2, i2);
  }
  return u2.__e;
}
function $(l2, u2) {
  n.__c && n.__c(u2, l2), l2.some(function(u3) {
    try {
      l2 = u3.__h, u3.__h = [], l2.some(function(n2) {
        n2.call(u3);
      });
    } catch (l3) {
      n.__e(l3, u3.__v);
    }
  });
}
function j(n2, l2, u2, i2, t2, r2, o2, c2) {
  var s2, a2, v2, h2, y2, p2 = u2.props, d2 = l2.props;
  if (t2 = l2.type === \\"svg\\" || t2, r2 != null) {
    for (s2 = 0; s2 < r2.length; s2++)
      if ((a2 = r2[s2]) != null && ((l2.type === null ? a2.nodeType === 3 : a2.localName === l2.type) || n2 == a2)) {
        n2 = a2, r2[s2] = null;
        break;
      }
  }
  if (n2 == null) {
    if (l2.type === null)
      return document.createTextNode(d2);
    n2 = t2 ? document.createElementNS(\\"http://www.w3.org/2000/svg\\", l2.type) : document.createElement(l2.type, d2.is && {is: d2.is}), r2 = null, c2 = false;
  }
  if (l2.type === null)
    p2 !== d2 && n2.data != d2 && (n2.data = d2);
  else {
    if (r2 != null && (r2 = e.slice.call(n2.childNodes)), v2 = (p2 = u2.props || f).dangerouslySetInnerHTML, h2 = d2.dangerouslySetInnerHTML, !c2) {
      if (r2 != null)
        for (p2 = {}, y2 = 0; y2 < n2.attributes.length; y2++)
          p2[n2.attributes[y2].name] = n2.attributes[y2].value;
      (h2 || v2) && (h2 && v2 && h2.__html == v2.__html || (n2.innerHTML = h2 && h2.__html || \\"\\"));
    }
    A(n2, d2, p2, t2, c2), h2 ? l2.__k = [] : (s2 = l2.props.children, g(n2, Array.isArray(s2) ? s2 : [s2], l2, u2, i2, l2.type !== \\"foreignObject\\" && t2, r2, o2, f, c2)), c2 || (\\"value\\" in d2 && (s2 = d2.value) !== void 0 && s2 !== n2.value && C(n2, \\"value\\", s2, p2.value, false), \\"checked\\" in d2 && (s2 = d2.checked) !== void 0 && s2 !== n2.checked && C(n2, \\"checked\\", s2, p2.checked, false));
  }
  return n2;
}
function H(l2, u2, i2) {
  try {
    typeof l2 == \\"function\\" ? l2(u2) : l2.current = u2;
  } catch (l3) {
    n.__e(l3, i2);
  }
}
function I(l2, u2, i2) {
  var t2, r2, o2;
  if (n.unmount && n.unmount(l2), (t2 = l2.ref) && (t2.current && t2.current !== l2.__e || H(t2, null, u2)), i2 || typeof l2.type == \\"function\\" || (i2 = (r2 = l2.__e) != null), l2.__e = l2.__d = void 0, (t2 = l2.__c) != null) {
    if (t2.componentWillUnmount)
      try {
        t2.componentWillUnmount();
      } catch (l3) {
        n.__e(l3, u2);
      }
    t2.base = t2.__P = null;
  }
  if (t2 = l2.__k)
    for (o2 = 0; o2 < t2.length; o2++)
      t2[o2] && I(t2[o2], u2, i2);
  r2 != null && a(r2);
}
function L(n2, l2, u2) {
  return this.constructor(n2, u2);
}
n = {__e: function(n2, l2) {
  for (var u2, i2; l2 = l2.__; )
    if ((u2 = l2.__c) && !u2.__)
      try {
        if (u2.constructor && u2.constructor.getDerivedStateFromError != null && (i2 = true, u2.setState(u2.constructor.getDerivedStateFromError(n2))), u2.componentDidCatch != null && (i2 = true, u2.componentDidCatch(n2)), i2)
          return w(u2.__E = u2);
      } catch (l3) {
        n2 = l3;
      }
  throw n2;
}}, d.prototype.setState = function(n2, l2) {
  var u2;
  u2 = this.__s !== this.state ? this.__s : this.__s = s({}, this.state), typeof n2 == \\"function\\" && (n2 = n2(u2, this.props)), n2 && s(u2, n2), n2 != null && this.__v && (l2 && this.__h.push(l2), w(this));
}, d.prototype.forceUpdate = function(n2) {
  this.__v && (this.__e = true, n2 && this.__h.push(n2), w(this));
}, d.prototype.render = p, u = [], i = typeof Promise == \\"function\\" ? Promise.prototype.then.bind(Promise.resolve()) : setTimeout, m.__r = 0, o = 0;

var t$1, u$1, r, o$1 = 0, i$1 = [], c$1 = n.__r, f$1 = n.diffed, e$1 = n.__c, a$1 = n.unmount;
function v$1(t2, r2) {
  n.__h && n.__h(u$1, t2, o$1 || r2), o$1 = 0;
  var i2 = u$1.__H || (u$1.__H = {__: [], __h: []});
  return t2 >= i2.__.length && i2.__.push({}), i2.__[t2];
}
function y(r2, o2) {
  var i2 = v$1(t$1++, 3);
  !n.__s && j$1(i2.__H, o2) && (i2.__ = r2, i2.__H = o2, u$1.__H.__h.push(i2));
}
function q() {
  i$1.some(function(t2) {
    if (t2.__P)
      try {
        t2.__H.__h.forEach(b), t2.__H.__h.forEach(g$1), t2.__H.__h = [];
      } catch (u2) {
        return t2.__H.__h = [], n.__e(u2, t2.__v), true;
      }
  }), i$1 = [];
}
n.__r = function(n2) {
  c$1 && c$1(n2), t$1 = 0;
  var r2 = (u$1 = n2.__c).__H;
  r2 && (r2.__h.forEach(b), r2.__h.forEach(g$1), r2.__h = []);
}, n.diffed = function(t2) {
  f$1 && f$1(t2);
  var u2 = t2.__c;
  u2 && u2.__H && u2.__H.__h.length && (i$1.push(u2) !== 1 && r === n.requestAnimationFrame || ((r = n.requestAnimationFrame) || function(n2) {
    var t3, u3 = function() {
      clearTimeout(r2), x$1 && cancelAnimationFrame(t3), setTimeout(n2);
    }, r2 = setTimeout(u3, 100);
    x$1 && (t3 = requestAnimationFrame(u3));
  })(q));
}, n.__c = function(t2, u2) {
  u2.some(function(t3) {
    try {
      t3.__h.forEach(b), t3.__h = t3.__h.filter(function(n2) {
        return !n2.__ || g$1(n2);
      });
    } catch (r2) {
      u2.some(function(n2) {
        n2.__h && (n2.__h = []);
      }), u2 = [], n.__e(r2, t3.__v);
    }
  }), e$1 && e$1(t2, u2);
}, n.unmount = function(t2) {
  a$1 && a$1(t2);
  var u2 = t2.__c;
  if (u2 && u2.__H)
    try {
      u2.__H.__.forEach(b);
    } catch (t3) {
      n.__e(t3, u2.__v);
    }
};
var x$1 = typeof requestAnimationFrame == \\"function\\";
function b(n2) {
  typeof n2.u == \\"function\\" && n2.u();
}
function g$1(n2) {
  n2.u = n2.__();
}
function j$1(n2, t2) {
  return !n2 || t2.some(function(t3, u2) {
    return t3 !== n2[u2];
  });
}

function App() {
      y(() => {
        console.log(\\"effect\\");
      }, []);
      return v(\\"div\\");
    }

    console.log(App);
"
`;

exports[`build with skypack 1`] = `
"var a,F,O;function I(t,e,n){var r,l,_,i=arguments,c={};for(_ in e)_==\\"key\\"?r=e[_]:_==\\"ref\\"?l=e[_]:c[_]=e[_];if(arguments.length>3)for(n=[n],_=3;_<arguments.length;_++)n.push(i[_]);if(n!=null&&(c.children=n),typeof t==\\"function\\"&&t.defaultProps!=null)for(_ in t.defaultProps)c[_]===void 0&&(c[_]=t.defaultProps[_]);return P(t,c,r,l,null)}function P(t,e,n,r,l){var _={type:t,props:e,key:n,ref:r,__k:null,__:null,__b:0,__e:null,__d:void 0,__c:null,__h:null,constructor:void 0,__v:l??++a.__v};return a.vnode!=null&&a.vnode(_),_}a={__e:function(t,e){for(var n,r,l,_=e.__h;e=e.__;)if((n=e.__c)&&!n.__)try{if((r=n.constructor)&&r.getDerivedStateFromError!=null&&(n.setState(r.getDerivedStateFromError(t)),l=n.__d),n.componentDidCatch!=null&&(n.componentDidCatch(t),l=n.__d),l)return e.__h=_,n.__E=n}catch(i){t=i;}throw t},__v:0},F=typeof Promise==\\"function\\"?Promise.prototype.then.bind(Promise.resolve()):setTimeout,O=0;

console.log(I);
"
`;
